src/Castle.Core/DynamicProxy/Internal/InvocationHelper.cs

```diff
@@ -48,22 +48,29 @@ public static MethodInfo GetMethodOnType(Type type, MethodInfo proxiedMethod)
 
 			Debug.Assert(proxiedMethod.DeclaringType.IsAssignableFrom(type),
 						 "proxiedMethod.DeclaringType.IsAssignableFrom(type)");
-			using (var locker = @lock.ForReadingUpgradeable())
+			using (var locker = @lock.ForReading())
```

Comment:
This is a great change, as you suspected this has got to be what the author of the code wanted. ProxyUtil was probably copy-and-pasted from this code and inherited the same flaw.